### ISSUE TITLE
Child extends Parent dependecy not working

### ISSUE THREAD

**Issue by cajoy** [angular/angular#5155]
I have parent with injected http service:

```ts
import {Inject, Injectable, Component} from 'angular2/angular2';
import {Http} from 'angular2/http';

@Injectable()

export abstract class BaseResource {
    http:   Http;

    constructor(@Inject(Http) http: Http) {
        console.log("INITIALIZED", http); /// Output INITIALIZED but http is undefined

        this.http = http;
    }
}
```

And child what extends from Parent:

```ts
import {Injectable, Component} from 'angular2/angular2';
import {Headers} from 'angular2/http';
import {BaseResource} from './BaseResource';

@Component({
    providers[BaseResource]
})

@Injectable()

export class CarModel extends BaseResource  {
    postData(carId: number) {
        this.http.post(this.config.apiUrl); // http is undefined 
    }
}
```

Please note that child has no constructor as it suppose to use parent dependencies.  But actually parent dependencies is not initalized for extends...

But it is ok if inject it directly without extends

**Comment by thelgevold** [comment]
I think in TypeScript you have to add super() in a child constructor
```
export class CarModel extends BaseResource{
constructor(){
      super();
    }

postData(carId: number) {
        this.http.post(this.config.apiUrl); // http is undefined 
    }
}
```

**Comment by cajoy** [comment]
Yes but in this case I need to inject http also what makes no sense...

Anyway parent constructor calls successfully but without injection

**Comment by pkozlowski-opensource** [comment]
@cajoy if you try to do this:

```javascript
class Foo {
  constructor(public bar) {};
}

class ExtendsFoo extends Foo {
	constructor() {};
}
```

You will see that this is not a valid TS. On the other hand doing:

```javascript
class ExtendsFoo extends Foo {
}
```

will transpile to:

```javascript
var ExtendsFoo = (function (_super) {
    __extends(ExtendsFoo, _super);
    function ExtendsFoo() {
        _super.apply(this, arguments);
    }
    return ExtendsFoo;
})(Foo);
```

which basically calls super's constructor with all arguments being undefined.

I hope that you can see from the above that you need to re-define and re-inject constructor arguments in a child class.

**Comment by cajoy** [comment]
Yes but in this case it makes no sense of e tended class as I need to redefine constructor parameters . It should be picked automatically and works fine for native TS but not works for DI

Is this is issue with DI or TS?

**Comment by pkozlowski-opensource** [comment]
> Yes but in this case it makes no sense of e tended class as I need to redefine constructor parameters  

I think this is an issue with what you expect. Let's take this example:

```javascript
class ExtendsFoo extends Foo {
    constructor(undefined) {};
}
```

What should happen here? Would you still expect DI to kick in? 

> Is this is issue with DI or TS?

The only "issue" I can see here is that TS doesn't report errors for the 

```javascript
class ExtendsFoo extends Foo {
}
```

case - you might want to open a ticket for this since IMO it should be the same as 

```javascript
class Foo {
  constructor(public bar) {};
}

class ExtendsFoo extends Foo {
	constructor() {};
}
```

**Comment by cajoy** [comment]
Let's assume I have following class where:

```ts
import {Http} from 'angular2/http';
import {Config} from '../../Config';

export abstract class BaseResource {
    config: Config;
    http:   Http;

    constructor() {
        this.http = new Http();
        this.config = new Config();
    }
}
```

Config is base TS class and is injected without any issues and I can use it's methods and properties.
While Http is angular DI injection class... So I can not use this component without proper injection.

So the issue I see is Angular2/Http  is not self-sustained component. So I can not use it as standalone.

It looks like Angular2 components are coupled to each other.

Is there any other way how I can inject Http in parent class? without touching Child?

**Comment by pkozlowski-opensource** [comment]
It is not about angular or DI - it is about presence of a constructor with arguments. If you extend a class that has a constructor with arguments you _need_ to do something about it - that is - call `super` with appropriate arguments. This is a requirement in any programming language I know of: with our without angular / di.

You seem to expect that ng's DI system would violate this basic principle by going over class hierarchy and magically invoke child constructors, which I don't think is a reasonable expectations. 

> It looks like Angular2 components are coupled to each other.

Once again, this has nothing to do with Angular. If one class extends another they are "coupled".

**Comment by cajoy** [comment]
So then.. Probably I don't have enough skills. But please help me to inject/instantiate http class in parent component.

I can easily use some kind of Restful.js , inject it in constructor and use it without any issue. But I would like to use native Angular2 /http component..

**Comment by woppa684** [comment]
You have to use the injector yourseld then (instead of letting angular do it).

**Comment by pkozlowski-opensource** [comment]
> But please help me to inject/instantiate http class in parent component.

```
export class CarModel extends BaseResource  {

    constructor(http: Http) {
         super(http);
    }

    ...
}

**Comment by cajoy** [comment]
@pkozlowski-opensource, Like I said, I want to avoid of declaring http component in parent

**Comment by pkozlowski-opensource** [comment]
> Like I said, I want to avoid of declaring http component in parent

Then your only option is to drop DI in your child component:

```ts
import {Http} from 'angular2/http';

export abstract class BaseResource {
    http = new Http();
}
```

But be aware that it voids all the benefits of DI

**Comment by cajoy** [comment]
@woppa684

```ts
import {Http} from 'angular2/http';
import {Config} from '../../Config';
import { Injector } from 'angular2/angular2';

export abstract class BaseResource {
    config: Config;
    http:   Http;

    constructor() {

        var injector = Injector.resolveAndCreateChild([
            Http
        ]);

        this.http = injector.get(Http);
        this.config = new Config();
    }
}
```

Gives me exception:

```
ORIGINAL EXCEPTION: No provider for ConnectionBackend! (Http -> ConnectionBackend)
```

**Comment by todoubaba** [comment]
@cajoy 
```typescript
let injector = Injector.resolveAndCreate([HTTP_PROVIDERS]);
this.http = injector.get(Http);
```

**Comment by cajoy** [comment]
@todoubaba Thanks a lot.

It did resolved all issues!...

**Comment by pkozlowski-opensource** [comment]
> It did resolved all issues!...

He he - I don't think so :-) Now you've got just a fancy way of doing `new Http()` with all the benefits of DI gone - but you have been warned :-)

**Comment by cajoy** [comment]
Yes. But let's assume I have 100 models for complicated application with couple of levels of inheritance. If I change constructor or added new service/provider - I will need to update all 100 constructors. While declaring in constructor - I will not need to do anything and just use new functionality

**Comment by cajoy** [comment]
There is good explanation of the issue: http://stackoverflow.com/questions/29012830/angular-js-di-with-es6-classes-and-inheritance

And it looks like there is solution like ...parentDependecies decorator what would be good to have as part of angular2 core

**Comment by gbataille** [comment]
Hi @pkozlowski-opensource 
So I understand your point. I can see how it would be complicated for the DI to intercept the chain of constructors in a chain of classes and superclasses.

Now @cajoy has a point. As far as I understand, Angular2 has no interceptors or filters (whatever you call them). But in this case, __what is the canonical way to address cross-cutting concerns?__
Coming readily to mind, I can mention
* a logger that would leave a trace on ALL component's ngOnInit hook
* an authentication brick that would redirect to the login page when you try to access a given protected component.

In Java, you would use AOP. In ruby, you would put a before_action (filter on the HTTP stack) in your top parent controller.

What about Angular2. The only things I have found (like @cajoy I guess) is
* doing a super class, but that needs some parameters (like the router). Which means any component extending it have to handle the super constructor (a bit cumbersome)
* doing an attribute directive, but then you have to add it to all your templates (like [here](http://stackoverflow.com/questions/34331478/angular2-redirect-to-login-page)

Can you share some insights please. Am I missing something, or am I "stuck" with the 2 mentioned solutions

Thanks

**Comment by zoechi** [comment]
Maybe https://aspect.codeplex.com/

**Comment by gbataille** [comment]
@zoechi interesting. You would rely on the Typescript compiler though and not on the framework. I guess that works. Would that be recommended though? is there any discussion, arguments on the subject in the angular2 community.

I'm sure this is a problem for everyone (or will be once Angular2 gains more speed). Will the community have a canonical answer?

**Comment by zoechi** [comment]
@gbataille cross-cutting concerns are relevant but "everyone" is probably a bit exagerated in the web-development area ;-). I'm interested in this topic but the priority isn't too high. I'm using Dart not TS, therefore TS solutions won't work for me anyway.
Currently if I want something global, I inject a service and call its methods.
In Dart I could build a transformer that probably would work similar like what the aspect tool linked above does. That doesn't involve Angular.
I haven't seen other discussions about this topic yet in the Angular2 community but I just started a few weeks ago.

**Comment by gbataille** [comment]
@zoechi yes I agree. I don't generally like magic (I got recently frustrated with Ruby dynamic method generation that is hard to follow). Without filters or AOP like pattens, the good thing is that everything is "clean" (understand the wiring is explicit and easy to follow)
It just that it feels a bit old to have to do something in all your components (like constructor argument passing) and I was really wondering if I was just not simply missing something.
But that possibly is the clean way to do it. I must say I'm not set myself. I just wanted to get a feel of what people might think about that :)

In any case, I'll stick to my constructor explicit pattern to build my app for now!

**Comment by zoechi** [comment]
Constructing new class instances is quite difficult with lots of pitfalls, therefore constructors are more cumbersome than other parts of an application.

**Comment by tbragaf** [comment]
I came accross the same issue.
The best solution is probably go with composition over inheritance.

Your before abstract class is now a "helper" which you will use to, let's say, build the request and send the request.

Something like:
```js 
class RequestManager {
  constructor(http: Http) { }
  /* other methods */
}

class Service {
  constructor(requestManager: RequestManager) {}
  /* other methods */
}
```

Does this make sense?

**Comment by gbataille** [comment]
Yup. That's what authHttp is in angular-jwt lib. But it's not exactly what
I needed for "redirect if not logged" kind of logic.

**Comment by tibbus** [comment]
When you use DI in a component you expect Angular2 to initialize the Class and call the constructor with the right params.
So I agree with @cajoy that in case of extending a Class you would expect that Angular2 also will call the parent constructor with the params for DI, but I also see how this thing will not work with Typescript.

 I will just inject the providers again in the child, it is how it is....

**Comment by mateusleon** [comment]
One way to do this, maybe, is to extend the `<router-outlet>` directive, which has an method called `activate`. When extending it, you can make you own logic and, if there's possibility to activate the route, you can call the super method activate, passing the same instruction to it.

```js
import {
    Attribute,
    Directive,
    DynamicComponentLoader,
    ViewContainerRef
} from 'angular2/core';

import {
    ComponentInstruction,
    Router,
    RouterOutlet
} from 'angular2/router';

import {
    Authentication
} from '../../providers/authentication';

@Directive({
    selector: 'router-outlet'
})
export class AuthenticationRouter extends RouterOutlet {
    constructor(
        viewRef: ViewContainerRef,
        loader: DynamicComponentLoader,
        parentRouter: Router,
        private authentication: Authentication,
        @Attribute('name') nameAttr: string
    ) {
        super(viewRef, loader, parentRouter, nameAttr);

        this.authentication.initSession();
    }

    activate(instruction: ComponentInstruction) {
        if (this._isPublic(instruction)) {
            return super.activate(instruction);
        }

        if (this.authentication.isLogged()) {
            return super.activate(instruction);
        }

        this.authentication.signIn();
    }

    private _isPublic(instruction: ComponentInstruction) {
        return instruction.routeData.data['authenticate'] === 'PUBLIC';
    }
}
```

That was my shoot on it. One thing to notice is that the session is initiated on this extension of `<router-outlet>` because of some trial and errors that I made, to find in which way this method where more effective. Initiating it on the constructor of this extension process my logic to find the authentication data on localStorage. If isn't found or it's expired (using oAuth '{token_type} {access_token}'), I redirect the user to another resource (remote) which makes the authentication. It returns then to the app, passing on the URI hash the params, which I parse on an specific route destinated to run the parser.

The `_isPublic()` method on this snippet is responsible to identify which is the type of the route, which is declared inside the Route/AsyncRoute `data` property.

Obs.: I'm still using the deprecated router (although this message has been sent when angular 2-rc.1 already had been released). I'm planning to change the code after the first stable release.

**Comment by richdyang** [comment]
composition over inheritance is better.

**Comment by nkagarw** [comment]
This exact scenario - Injecting services in the base class constructor and accessing those in a derived class with no constructor is working for me with Angular 2.4.10 and TypeScript 2.1.4. It stops working as soon as I upgrade TS to 2.1.6 or 2.2.x.

**Comment by DzmitryShylovich** [comment]
@nkagarw reproduce in [plunkr](http://plnkr.co/edit/lU32vBJY9KYXZUTW7GQQ?p=preview) and file a new isssue

**Comment by nkagarw** [comment]
@DzmitryShylovich here is a plunker : [http://plnkr.co/edit/y7sQdv4lXHU6Q34GlWMw?p=preview]

The AppComponent has no constructor whereas BaseComponent (its parent) has a constructor with arguments (a small service injected). Properties from the BaseComponent are accessible in AppComponent. It is working well as long as I keep the Typescript version to 2.1.4. It breaks as soon as I change to TS v2.1.6.

**Comment by DzmitryShylovich** [comment]
@nkagarw angular 2.x doesn't support ts 2.1+
the latest supported version is 2.0.10
if you want to use ts 2.1+ you need angular 4.x

**Comment by MartinFreire** [comment]
Hi, I know this is an old issue but I faced the same problem.
My solution to solve this was to inject a service that has methods to every argument needed on the super constructor.
`super(parentSvc.getParam1(),parentSvc.getParam2())`
This way I can change the type of the param1 on the class and the service only, the childs doesn't need to know wath is beeing injected, just the number of params
So if my parent class constructor is 
`constructor( prop1: MyFooSvc, prop2: MyOtherFooSvc)`
My parentSvc getParam1() will be like this:
`getParam1(): MyFooSvc`
And I could change it without having to update every child
Is this a correct way of doing it?
Thanks!

**Comment by nkagarw** [comment]
@DzmitryShylovich Thanks for the info. I upgraded the [plunker](http://plnkr.co/edit/I7l2f53EGx0b26vLUtXm?p=preview) to use Angular 4 with TS2.1 and it works great. 

Actually, the point I was trying to make is that it is possible to extend a base class that has parameterized constructor with a constructor-less child class. FYI for @cajoy and @pkozlowski-opensource.

**Comment by dalbir** [comment]
Here is the solution 👍 

https://stackoverflow.com/a/40592524/2290351

**Comment by abelkbil** [comment]
@cajoy and @todoubaba Thanks for insight.

Angular 4.0
My base service class for api is as below
```
import { ReflectiveInjector } from '@angular/core';
import { Headers, Http, HttpModule, RequestOptions, Response } from '@angular/http';
import { Observable } from 'rxjs/Observable';

export  class BaseService {

  http:   Http;

  constructor() {
    let injector = ReflectiveInjector.resolveAndCreate([HttpModule, Http]);
    this.http = injector.get(Http);
  }

  get(url) {
    return this.http.get(url).map(res => res.json()).catch(this.handleError);
  }

  delete(url) {
    return this.http.delete(url).map(res => res.json()).catch(this.handleError);
  }

  extractData(res: Response) {
    const body = res.json();
    return body || {};
  }

  handleError(error: any) {
    const errMsg = (error.message) ? error.message :
        error.status ? `${error.status} - ${error.statusText}` : 'Server error';
    return Observable.throw(errMsg);
  }
}
```

Try added 

> HttpModule, Http
even though it throws No provider for ConnectionBackend! (Http -> ConnectionBackend)
